### PR TITLE
TypeScript rules

### PR DESCRIPTION
**Pull request by molant**
## Pull request checklist

Make sure you:

- [X] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [X] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [X] Added/Updated related documentation.
- [X] Added/Updated related tests.

## Short description of the change(s)

This PR adds a couple new rules for TypeScript. Couple things that I'd like to do first this PR is merged:

* Find a better name for the `target` rule
* Complete the `target` support matrix, `browserslist` returns a lot of browsers in some of the queries and this should be more complete
* Maybe add what browser(s) are affecting the `target`? 

In any case, this is ready to start looking in case there's any major change required.

### CODE REVIEW

**Comment by molant** [comment]
This PR also fixes some stuff that could be merged before (first 3 commits)

**Comment by alrra** [comment]
> This PR also fixes some stuff that could be merged before (first 3 commits)

Done.

File: packages/rule-typescript-config/src/target.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+         * account ranges and minor versions. E.g.: `4.4.3-4.4.4` will use `4.4.3` to compare
+         */
+        const isOlder = (version1: number | string, version2: number | string): boolean => {
+
+            if (typeof version1 === 'number' && typeof version2 === 'number') {
+                return version1 < version2;
+            }
+
+            // Can't store `typeof` or otherwise TypeScript complains about `includes` not being on `number`
+            if (typeof version1 === 'string' && typeof version2 === 'string') {
```
Comment (sarvaje) [inline packages/rule-typescript-config/src/target.ts]: IMO, this is to complex (or I'm missing something). Why don't you just parse the versions?
```
if(typeof v1 === 'string') {
v1 = getMajor(v1)
}
if(typeof v2 === 'string') {
v2 = getMajor(v2)
}
return isOlder(v1, v2)
```

File: packages/rule-typescript-config/src/target.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            const versions = Object.keys(compatMatrix);
+            let maxVersion = 'es3';
+
+            /**
+             * This will check all the ES versions and compare all the
+             * browsers from the compat matrix agains the given ones.
+             * If there's one in the compat matrix newer than the ones
+             * provided, it will fail.
+             */
+            versions.forEach((version) => {
```
Comment (sarvaje) [inline packages/rule-typescript-config/src/target.ts]: shouldn't  this be a `reduce`?

File: packages/rule-typescript-config/src/target.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        };
+
+        /**
+         * Returns the configured normalized target:
+         *
+         * * The `es` part will be upper cased (e.g.: ES2015, ESNext)
+         * * ES6 --> ES2015
+         */
+        const getConfiguredTarget = (config: TypeScriptConfig): string => {
+            const target = config && config.compilerOptions && config.compilerOptions.target || 'es3';
```
Comment (sarvaje) [inline packages/rule-typescript-config/src/target.ts]: I think you don't need to check if `config` exists, if you receive an `parse::typescript-config` then the config has to exist.

File: packages/rule-typescript-config/src/target.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+            return Targets.get(target.toLowerCase());
```
Comment (sarvaje) [inline packages/rule-typescript-config/src/target.ts]: `Target` is not necessary, you can return `target.toLowerCase()` directly

File: packages/rule-typescript-config/README.md [hunk 5]
```diff
@@ -50,9 +62,171 @@ The configuration is valid:
...
+issues with your users.
+
+### What does the rule check?
+
+This rule checks the `compilerOptions` property `target` is the appropriate based
+on your `browserslist` configuration.
+
+The matrix is as follows:
+
+| Target | Chrome | Edge | Firefox | IE | Safari |
```
[the remark this answers is not in the document]
Comment (molant) [reply packages/rule-typescript-config/README.md]: Nope, see the comment in `Short description of the change(s)`

File: packages/rule-typescript-config/src/target.ts (hunk 4 above)
Comment (molant) [reply packages/rule-typescript-config/src/target.ts]: target could be `es6` and I need to transform it to `es2015` (thus my issue about normalizing things)

File: packages/rule-typescript-config/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/rule-typescript-config/README.md]: ups, I missed that

File: packages/rule-typescript-config/src/target.ts (hunk 4 above)
Comment (sarvaje) [reply packages/rule-typescript-config/src/target.ts]: You are right, I missed that value.

File: packages/rule-typescript-config/src/target.ts (hunk 1 above)
Comment (molant) [reply packages/rule-typescript-config/src/target.ts]: Browser versions are complicated...
We could be comparing:
* `safari 10.1` and `safari 10.0`
* `android 3.0.1` and `android 4.4.3`
* In some cases we get a range as a version: `android 4.4.3-4.4.4`.

I should put some comments or try to refactor the code to make it clearer.

File: packages/rule-typescript-config/src/target.ts (hunk 2 above)
Comment (molant) [reply packages/rule-typescript-config/src/target.ts]: Yes, I can probably change it to `reduce`

File: packages/rule-typescript-config/src/target.ts (hunk 1 above)
Comment (sarvaje) [reply packages/rule-typescript-config/src/target.ts]: but at the end, you just compare the major version, isn't it?

Comment (molant) [reply packages/rule-typescript-config/src/target.ts]: I should compare the mayor only if the other tests have failed.

Comment (sarvaje) [reply packages/rule-typescript-config/src/target.ts]: ah, ok

File: packages/rule-typescript-config/docs/target.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+This rule checks the `compilerOptions` property `target` is the appropriate based
+on your `browserslist` configuration.
+
+The matrix is as follows:
+
+| Target | Chrome | Edge | Firefox | IE | Safari |
+| ------ | ------ | ---- | ------- | -- | ------ |
+|  ES3   |   ✔    |  ✔  |    ✔    | ✔ |    ✔   |
+|  ES5   |   57   |  12  |   45    | 9 |    8    |
```
Comment (molant) [inline packages/rule-typescript-config/docs/target.md]: I'm sure the values for chrome are not the right ones. Need to recheck the kangax tables and update this accordingly.

**Comment by molant** [comment]
@sarvaje @alrra, I've updated the browser support matrix and added a few more links. Let me know if you have any other feedback or otherwise I'll merge it as soon as I can.

**Pull request merged**